pkg/kv/kvserver/rangefeed_restarter.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package kvserver
+
+import (
+	"context"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+var rangefeedRestartInterval = settings.RegisterDurationSetting(
+	settings.SystemOnly,
+	"kv.rangefeed.restart_interval",
+	"minimum interval between rangefeed restarts to conform to new settings",
+	50*time.Millisecond,
+	settings.WithVisibility(settings.Reserved),
+)
+
+var rangefeedRestartAvailablePercent = settings.RegisterIntSetting(
+	settings.SystemOnly,
+	"kv.rangefeed.restart_quota_headroom",
+	"minimum percentage of rangefeed catchup quota that still allow processor "+
+		"reconfiguration restarts to proceed. set to >100 to disable eager restarts on "+
+		"processor type change",
+	50,
```

Comment:
I think there's an interaction here that's going to cause this to not throttle at all:

`kv.rangefeed.concurrent_catchup_iterators` defaults to 16 (server-side). `kv.rangefeed.catchup_scan_concurrency` defaults to 8 (client-side). The value of 50% here is inclusive, and applies to the server-side, so it will keep restarting rangefeeds when there's 8 concurrent catchup scans on the server.

Notice that the client will throttle catchup scans at 8. Consider the common case where the client is colocated with the server via DistSQL, and there's only 1 changefeed running on a range. The client will never start more than 8 catchup scans concurrently, so the restarter will continue to restart all rangefeeds, which will queue on the client.

This is not an issue if we remove the client-side semaphore, as we've discussed. But if we merge this as-is, we have to make sure we do that before release. Alternatively, we can take the minimum of the client-side and server-side semaphore sizes and apply the fraction to that.


Also, nit: I think we typically use fractions rather than percentages in settings, i.e. 0.5, but haven't checked.

Reply:
Getting client semaphore configuration doesn't make much sense. What if we set default fraction to 1? That would mean we won't try to restart if there's at least one catch up happening.
Current client side limiter avoids spawning all requests at once, can we really remove that?

Reply:
> we won't try to restart if there's at least one catch up happening.

Won't it take forever to restart e.g. 1 million ranges in that case?

> Current client side limiter avoids spawning all requests at once, can we really remove that?

We can rate limit it instead. It's primarily intended to avoid spawning too many goroutines too fast, which we can also do simply by sleeping between each spawn.